src/main/java/de/zalando/aruha/nakadi/service/EventStream.java

```diff
@@ -55,23 +88,37 @@ public void streamEvents() {
             int messagesRead = 0;
             int messagesReadInBatch = 0;
 
-            Map<String, List<String>> currentBatch = emptyBatch();
-            final Map<String, String> latestOffsets = Maps.newHashMap(config.getCursors());
+            Map<TopicPartition, List<String>> currentBatch = emptyBatch();
+            Map<TopicPartition, String> latestOffsets = cursorsToOffsetMap(initialCursors);
 
             long start = currentTimeMillis();
             long batchStartTime = start;
 
             while (true) {
+
+                // check if rebalance is invoked
+                final Optional<List<Cursor>> newCursorsOrNone = newDistribution.getAndSet(Optional.empty());
+                if (newCursorsOrNone.isPresent()) {
+                    final List<Cursor> newCursors = newCursorsOrNone.get();
+                    eventConsumer.setCursors(newCursors);
+                    partitions = cursorsToPartitions(newCursors);
+                    latestOffsets = cursorsToOffsetMap(newCursors);
+                    // forget about all messages not flushed
+                    messagesReadInBatch = 0;
+                    currentBatch = emptyBatch();
+                    batchStartTime = currentTimeMillis();
+                }
+
                 final Optional<ConsumedEvent> eventOrEmpty = eventConsumer.readEvent();
```

Comment:
I again examined the possibility to implement that. And now I think that I would not like to do that at least for now. The long polls (when poll time equals the batch timeout) can bring problems. One of them is if we decide to switch to per-partition batch-timeout tracking - then the long polls will not make sense. (And as I understand from some discussion in the past - switching to per-partition timeouts is quite possible)
So for now I would like to leave this part as it is. (And maybe implement it later as an improvement)